### DISCUSSION TITLE
Regarding the version numbering scheme

### DISCUSSION THREAD

**Discussion by dabnciencias** [ManimCommunity/manim#1764] [General]
I understand a decision has been made to release a new ManimCommunity version each month to keep everyone participating in its development motivated, which seems great. However, the current numbering scheme being time based is unideal, since the quantity and quality of improvements made in a month's time can vary quite arbitrarily. Therefore, I'd suggest changing to a more useful numbering scheme, which is more indicative of both the amount and depth of the changes made in each release. This could help people currently doing animation projects in manim to have a rough idea of the changes made and whether or not to consider upgrading to the latest version, and not feel 'lost' regarding the newest version with each coming month -as has personallly happened to me and the team I work with.

**Comment by behackl** [comment]
Could you give an example of what exactly you mean with a "more useful numbering scheme"?

Ultimately, we are releasing often in order to "keep things flowing" and to distribute the bugfixes and enhancements made during one release cycle to users. (Or at least: to give them the opportunity to upgrade to a newer version.) In that regard, we are putting quite some effort towards maintaining a [changelog](https://docs.manim.community/en/stable/changelog/0.8.0-changelog.html) giving an overview over all the changes we have made within a release cycle.

Now, from our side it would be easy for releases like the current one, where we are – at the time of releasing – not aware of any breaking changes introduced to the library, to make it a minor release (i.e., `v0.7.1` instead of `v0.8.0`). We can definitely pay more attention towards this in the future, but I don't think that we want to change the monthly releases; at least not for the next few ones.

If a particular version of the community edition is already working for you and your team, I'd recommend to simply stick with that version for a given project – but if you upgrade and something is broken that isn't mentioned in a "Breaking Changes" section in our changelog, please definitely do file a bug report: chances that we don't discover it on our own are, unfortunately, pretty large.

**Comment by WampyCakes** [comment]
I don't think it's fair to say that our numbering scheme is time based. Our _release schedule_ is time based, but our versioning is not. Aside from keeping development pumping along, since we're still at a _somewhat_ unstable point in the project, time based releases are the best way to conveniently allow people to experience the newest improvements in a timely, stable manner. 

To the best that I can tell and from what I vaguely remembering being decided before, we're loosely following semantic versioning. [This](https://gist.github.com/jashkenas/cbd2b088e20279ae2c8e) is a pretty good argument against SemVer in my opinion. I really agree with what he said here: 
> For a long time in the history of software, version numbers indicated the relative progress and change in a given piece of software. A major release (1.x.x) was major, a minor release (x.1.x) was minor, and a patch release was just a small patch. You could evaluate a given piece of software by name + version, and get a feeling for how far away version 2.0.1 was from version 2.8.0.

Now, I actually think we're following this ideal of major, minor, and patch pretty well. We have not reached a point where we have a major release -- we still have some big milestones ahead of a major release that we would like to tackle. We also aren't churning out patch releases (v0.1.1 was kind of an anomaly since we were still early in the project and figuring out how we would release versions). So I think we're well within releasing these, more or less, as minor releases. I could see here an argument for nightly releases or a `vW.X.Y.Z` scheme, but practically this won't work well with PyPI to the best of my knowledge.

If we started releasing our monthly releases as patch releases (like `v0.7.1`), then the meaning of the third number is lost. It no longer indicates that it is a very small update where pretty much the only change is a bugfix (which has been repeatedly discussed as a backup option if we merge something before a monthly release that ends up breaking important functionality). That all being said, if we were to change our release schedule or versioning scheme, I think it would be best to do so after a major release and not in the middle of major releases.

> since the quantity and quality of improvements made in a month's time can vary quite arbitrarily.

While this can be true (and probably is more true of this last release with people being busy), I would argue that our monthly releases are, more or less, pretty consistent with each other. I think they belong as minor releases for now, and I think they generally consistently come with a lot of enhancements, bugfixes, and new features. 

> Therefore, I'd suggest changing to a more useful numbering scheme, which is more indicative of both the amount and depth of the changes made in each release ... This could help people currently doing animation projects in manim to have a rough idea of the changes made and whether or not to consider upgrading to the latest version, and not feel 'lost' regarding the newest version with each coming month -as has personallly happened to me and the team I work with.

As behackl said, read the changelog! It's there for your benefit so that you can discern for yourself whether or not you should/need to upgrade as well as to see what's changed. Just because there is a new release does not mean you must upgrade, and if you're in the middle of a project, I would never upgrade unless I _absolutely need_ something from the new release. If you do find yourself needing something, check the PR mentioned in the changelog for that change and see if it's a change that you can easily replicate yourself locally. Sometimes replicating a change to fix a bug or add a feature is better than upgrading -- particularly if you're in the middle of development.

**Reply by WampyCakes** [reply]
Perhaps at some point we should look into some sort of nighty(/weekly/other time period?) releases that can be installed with pip from GitHub or something of this sort. Though I imagine our actual versioning scheme won't change much from this. It would be more to give people the latest changes easily and would be a bit more conducive imo to developing for longer periods than a month to focus on big features, reworks, and more major release type material.

**Reply by naveen521kk** [reply]
You can always install the main branch using `pip install https://github.com/ManimCommunity/manim/archive/refs/heads/main.zip`

**Reply by WampyCakes** [reply]
You can and I was thinking of something similar where we have a script autogenerate GitHub releases nightly (or some fixed period of time). The benefit to this instead of installing main is that it gives a better point of reference to anyone trying to help someone who installed a nightly build.

**Reply by Darylgolden** [reply]
Is there another library that does something similar?

**Reply by WampyCakes** [reply]
Nightly releases aren't uncommon. For example, Mozilla uses them as well. Here's an example https://ftp.mozilla.org/pub/firefox/nightly/

**Comment by kilacoda-old** [comment]
Moved this to a discussion as it seemed more appropriate. Pinging @dabnciencias @WampyCakes @behackl  in case you didn't get a notif for it.
